streams/jvm/src/main/scala/zio/stream/platform.scala

```diff
@@ -537,28 +546,37 @@ trait ZStreamPlatformSpecificConstructors {
      * Write the entire Chuck[Byte] to the socket channel.
      * The caller of this function is NOT responsible for closing the `AsynchronousSocketChannel`.
      *
-     * The sink will yield the count of bytes written.
+     * The sink will yield true if stream was completely written.
      */
-    def write: Sink[Throwable, Byte, Nothing, Int] =
-      ZSink.foldLeftChunksM(0) { case (nbBytesWritten, c) =>
-        IO.effectAsync[Throwable, Int] { callback =>
-          socket.write(
-            ByteBuffer.wrap(c.toArray),
-            null,
-            new CompletionHandler[Integer, Void] {
-              override def completed(result: Integer, attachment: Void): Unit =
-                callback(ZIO.succeed(nbBytesWritten + result.toInt))
-
-              override def failed(error: Throwable, attachment: Void): Unit = callback(ZIO.fail(error))
-            }
-          )
+    def write: Sink[Throwable, Byte, Byte, Boolean] =
+      ZSink.foldChunksM(true)(Predef.identity) {
+        case (_: Boolean, c: Chunk[Byte]) => {
+          val buffer = ByteBuffer.wrap(c.toArray)
+          IO.effectAsync[Throwable, Boolean] { callback =>
+            socket.write(
+              buffer,
+              buffer,
+              new CompletionHandler[Integer, ByteBuffer] {
+                override def completed(result: Integer, attachment: ByteBuffer): Unit =
+                  if (attachment.hasRemaining)
+                    socket.write(attachment, attachment, this)
+                  else
+                    callback(ZIO.succeed(true))
+
+                override def failed(error: Throwable, attachment: ByteBuffer): Unit = error match {
+                  case _: ClosedChannelException => callback(ZIO.succeed(false))
+                  case _                         => callback(ZIO.fail(error))
+                }
+              }
+            )
+          }
         }
       }
 
     /**
      * Close the underlying socket
      */
-    def close(): UIO[Unit] = ZIO.effectTotal(socket.close())
+    def close(): UIO[Unit] = ZIO.effect(socket.close()).ignore
```

Comment:
This is a bit problematic as we'd want exceptions thrown here to be turned into defects.

Reply:
My rationale:
1. Network connection is impacted by remote systems and other events completely out of our control.
There is nothing we can do to resolve and an execution trace referencing them has no value.
2. A close() indicates we have no further interest in this connection. 
That would include whether it had trouble in closing. 

I did see that `effectTotal(close())` is the standard approach in the code base.
That approach assumes that exceptions will be very rarely thrown and are not of interest to the programmer.

Seems that is not appropriate when writing to Files.
A failure on close is quite likely (disk quote exceeded, NAS offline, etc).
The client code can sensibly recover in many cases, e.g. by recreating the file in a different location. 
That would require the file close use an error rather than a defect.

I am happy to change the code back to effectTotal.